src/dune_rules/dune_env.ml

```diff
@@ -19,7 +19,31 @@ module Stanza = struct
       Option.map since ~f:(fun since ->
           Dune_lang.Syntax.since Menhir_stanza.syntax since)
     in
-    Ordered_set_lang.Unexpanded.field "menhir_flags" ?check
+    let* menhir_version = Dune_lang.Syntax.get Menhir_stanza.syntax in
+    let+ flags = Ordered_set_lang.Unexpanded.field "menhir_flags" ?check in
+    Ordered_set_lang.Unexpanded.fold_strings flags ~init:()
+      ~f:(fun _pos sw acc ->
+        match String_with_vars.text_only sw with
+        | None -> acc
+        | Some text -> (
+          if String.equal text "--explain" then
+            match menhir_version with
+            | None -> ()
```

Comment:
We do nothing if getting the menhir lang failed for whatever reason. This is fine since this is just a check for the user. This logic will be cleaned up in a later PR.